lib/value/src/kind/comparison.rs

```diff
@@ -0,0 +1,480 @@
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_boolean(&self) -> bool {
+        self.boolean.is_some()
+    }
+
+    /// Returns `true` if the type is _at least_ `timestamp`.
+    ///
+    /// Note that other type states can also still be valid, for exact matching, also compare
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_timestamp(&self) -> bool {
+        self.timestamp.is_some()
+    }
+
+    /// Returns `true` if the type is _at least_ `regex`.
+    ///
+    /// Note that other type states can also still be valid, for exact matching, also compare
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_regex(&self) -> bool {
+        self.regex.is_some()
+    }
+
+    /// Returns `true` if the type is _at least_ `null`.
+    ///
+    /// Note that other type states can also still be valid, for exact matching, also compare
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_null(&self) -> bool {
+        self.null.is_some()
+    }
+
+    /// Returns `true` if the type is _at least_ `array`.
+    ///
+    /// Note that other type states can also still be valid, for exact matching, also compare
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_array(&self) -> bool {
+        self.array.is_some()
+    }
+
+    /// Returns `true` if the type is _at least_ `object`.
+    ///
+    /// Note that other type states can also still be valid, for exact matching, also compare
+    /// against `is_exact()`.
+    #[must_use]
+    pub fn is_object(&self) -> bool {
+        self.object.is_some()
+    }
+
+    /// Returns `true` if exactly one type is set.
+    ///
+    /// For example, the following:
+    ///
+    /// ```rust,ignore
+    /// kind.is_float() && kind.is_exact()
+    /// ```
+    ///
+    /// Returns `true` only if the type is exactly a float.
+    #[must_use]
+    #[allow(clippy::many_single_char_names)]
+    pub fn is_exact(&self) -> bool {
+        let a = self.is_bytes();
+        let b = self.is_integer();
+        if !(!a || !b) {
+            return false;
+        }
+
+        let c = self.is_float();
+        if !(!c || !a && !b) {
+            return false;
+        }
+
+        let d = self.is_boolean();
+        if !(!d || !a && !b && !c) {
+            return false;
+        }
+
+        let e = self.is_timestamp();
+        if !(!e || !a && !b && !c && !d) {
+            return false;
+        }
+
+        let f = self.is_regex();
+        if !(!f || !a && !b && !c && !d && !e) {
+            return false;
+        }
+
+        let g = self.is_null();
+        if !(!g || !a && !b && !c && !d && !e && !f) {
+            return false;
+        }
+
+        let h = self.is_array();
+        if !(!h || !a && !b && !c && !d && !e && !f && !g) {
+            return false;
+        }
+
+        let i = self.is_object();
+        if !(!i || !a && !b && !c && !d && !e && !f && !g && !h) {
+            return false;
+        }
+
```

Comment:
I think a slightly easier to follow, albeit marginally slower implementation would be:

```rust
let mut count = 0;
if self.is_bytes() {
  count += 1;
}
if self.is_integer(){
  count += 1;
}
if self.is_float(){
  count += 1;
}
if self.is_boolean(){
  count += 1;
}
if self.is_timestamp(){
  count += 1;
}
if self.is_regex() {
  count += 1;
}
if self.is_null(){
  count += 1;
}
if self.is_array(){
  count += 1;
}
if self.is_object(){
  count += 1;
}
count == 1
```

Reply:
Agreed. That is easier to reason about, and performance isn't that much of an issue in this code path.

Reply:
(Might also just use `self.is_bytes() + ... + self.is_object() == 1`.)

Reply:
`is_*` returns a boolean, not sure how that would work?

Reply:
Perhaps: 

```rust
let a = |b| if b { 1 } else { 0 };

a(self.is_bytes()) + 
a(self.is_integer()) + 
a(self.is_float()) + 
a(self.is_boolean()) + 
a(self.is_timestamp()) + 
a(self.is_regex()) + 
a(self.is_null()) + 
a(self.is_array()) + 
a(self.is_object()) == 1
```